File: core/src/main/java/dev/vibris/core/ResumeResponses.java
package dev.vibris.core;

import dev.vibris.core.request.RequestRegistry;
import dev.vibris.protocol.v1.ClientMessage;
import dev.vibris.protocol.v1.JobSummary;
import dev.vibris.protocol.v1.ResumeState;
import dev.vibris.protocol.v1.ServerMessage;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

final class ResumeResponses {
    private ResumeResponses() {
    }

    static List<ServerMessage> create(
        ControlSession session,
        ClientMessage message,
        RequestRegistry<TerminalResult> requests,
        Map<String, CoreJob> liveJobs
    ) {
        ResumeState.Builder result = ResumeState.newBuilder();
        List<ServerMessage> terminalReplay = new ArrayList<>();
        for (String requestId : message.getResumeRequest().getRequestIdsList()) {
            requests.resume(requestId, session.workspaceId()).ifPresent(snapshot -> {
                CoreJob job = liveJobs.get(requestId);
                if (job != null && job.workspaceId.equals(session.workspaceId())) job.bind(session);
                result.addJobs(JobSummary.newBuilder()
                    .setRequestId(requestId)
                    .setState(ProtocolMessages.jobState(snapshot.state())));
                if (snapshot.state().terminal()) {
                    terminalReplay.add(snapshot.result().message(
                        message.getMessageId(), requestId, session.workspaceId()));
                }
            });
        }
        List<ServerMessage> responses = new ArrayList<>(terminalReplay.size() + 1);
        responses.add(ProtocolMessages.envelope(
            message.getMessageId(), message.getRequestId(), session.workspaceId())
            .setResumeState(result)
            .build());
        responses.addAll(terminalReplay);
        return responses;
    }
}

File: core/src/main/kotlin/dev/vibris/core/ResumeResponses.kt
package dev.vibris.core

import dev.vibris.core.request.RequestRegistry
import dev.vibris.protocol.v2.ClientMessage
import dev.vibris.protocol.v2.ErrorCode
import dev.vibris.protocol.v2.JobStateSnapshot
import dev.vibris.protocol.v2.JobSummary
import dev.vibris.protocol.v2.ServerMessage

internal object ResumeResponses {
    @JvmStatic
    fun create(
        session: ControlSession,
        message: ClientMessage,
        requests: RequestRegistry<TerminalResult>,
        liveJobs: Map<String, CoreJob>,
    ): List<ServerMessage> {
        val requestId = message.resumeJob.jobId
        val snapshot = requests.resume(requestId, session.workspaceId()).orElse(null)
            ?: return listOf(
                ProtocolMessages.failure(
                    requestId,
                    requestId,
                    ErrorCode.ERROR_CODE_SERVER_RESTARTED,
                    "The requested job is not present in this Vibris server instance and may be submitted again.",
                ).message(message.messageId, message.requestId, session.workspaceId()),
            )
        snapshot.result?.let { terminal ->
            return listOf(terminal.message(message.messageId, message.requestId, session.workspaceId()))
        }
        val job = liveJobs[requestId]
        if (job != null && job.workspaceId == session.workspaceId()) {
            job.bind(session)
        }
        val summary = JobSummary.newBuilder()
            .setJobId(job?.submission?.jobId ?: requestId)
            .setRequestId(requestId)
            .setWorkspaceId(session.workspaceId())
            .setOperation(job?.submission?.workloadCase?.name?.lowercase().orEmpty())
            .setState(ProtocolMessages.jobState(snapshot.state))
            .build()
        val state = JobStateSnapshot.newBuilder().setSummary(summary)
        snapshot.result?.completed?.result?.let(state::setResult)
        snapshot.result?.failed?.error?.let(state::setError)
        return listOf(
            ProtocolMessages.envelope(message.messageId, message.requestId, session.workspaceId())
                .setJobState(state)
                .build(),
        )
    }
}
